README.md
################################################################################
 Laravel Breadcrumbs
################################################################################

A simple Laravel-style way to create breadcrumbs in `Laravel 5 <http://laravel.com/>`_.

Based on the the original [laravel-breadcrumbs](https://github.com/davejamesmiller/laravel-breadcrumbs) by Dave James Miller. 

Released under the MIT License.

###[Documentation](https://laravel-breadcrumbs.readthedocs.io/)

docs/changelog.rst
.. warning::

    Laravel Breadcrumbs is no longer maintained. Please see the `README <https://github.com/davejamesmiller/laravel-breadcrumbs/blob/master/README.rst>`_ for more details.

################################################################################
 Changelog
################################################################################

.. role:: date
    :class: changelog-date

.. role:: future
    :class: changelog-future

.. role:: strikethrough
    :class: strikethrough


Laravel Breadcrumbs uses `Semantic Versioning <http://semver.org/>`_.


.. ================================================================================
..  v3.0.1_ :future:`(Unreleased)`
.. ================================================================================


================================================================================
 v3.0.1_ :date:`(28 Aug 2016)`
================================================================================


- Laravel 5.3 support

.. _v3.0.1: https://github.com/davejamesmiller/laravel-breadcrumbs/tree/3.0.1


----------------------------------------
 Upgrading from Laravel 5.2 to 5.3
----------------------------------------

- Upgrade Laravel Breadcrumbs to 3.0.1 (or above)
- Move ``app/Http/breadcrumbs.php`` to ``routes/breadcrumbs.php`` (optional but recommended)


================================================================================
 v3.0.0_ :date:`(8 Feb 2015)`
================================================================================

- Add Laravel 5 support (`#62`_)
- Change view namespace from ``laravel-breadcrumbs::`` to ``breadcrumbs::``
- Change Bootstrap 3 template from ``<ul>`` to ``<ol>`` to match the `documentation <http://getbootstrap.com/components/#breadcrumbs>`_
- Move documentation from GitHub (Markdown) to `Read The Docs <https://readthedocs.org/>`_ (reStructuredText/`Sphinx <http://sphinx-doc.org/>`_)
- Greatly improve unit & integration tests (largely thanks to `Testbench <https://github.com/orchestral/testbench>`_)
- Fix issue that prevented non-deferred service providers referencing Breadcrumbs (`#39`_) by making Breadcrumbs non-deferred also
- Rename ``generateArrayIfExists()`` to ``generateIfExistsArray()``
- Rename ``renderArrayIfExists()`` to ``renderIfExistsArray()``
- Remove ``$breadcrumbs->get()`` and ``$breadcrumbs->set()`` methods from Generator class (they were never used nor documented)
- Remove ``Breadcrumbs::getView()``
- Switch from PSR-0 to PSR-4 file naming

.. _v3.0.0: https://github.com/davejamesmiller/laravel-breadcrumbs/tree/3.0.0
.. _#39: https://github.com/davejamesmiller/laravel-breadcrumbs/issues/39
.. _#62: https://github.com/davejamesmiller/laravel-breadcrumbs/issues/62


----------------------------------------
 Upgrading from 2.x to 3.x
----------------------------------------

- `Upgrade to Laravel 5 <http://laravel.com/docs/5.0/upgrade#upgrade-5.0>`_
- Move ``app/breadcrumbs.php`` to ``app/Http/breadcrumbs.php``
- Move ``app/config/packages/davejamesmiller/laravel-breadcrumbs/config.php`` to ``config/breadcrumbs.php`` (if used)

The following changes are optional because there are shims in place:

- In the config file, replace ``laravel-breadcrumbs::`` with ``breadcrumbs::``
- Replace any calls to ``Breadcrumbs::generateArrayIfExists()`` with ``Breadcrumbs::generateIfExistsArray()``
- Replace any calls to ``Breadcrumbs::renderArrayIfExists()`` with ``Breadcrumbs::renderIfExistsArray()``

.. note::

    Laravel 4 and PHP 5.3 are no longer supported -- please continue to use the `2.x branch <https://github.com/davejamesmiller/laravel-breadcrumbs/tree/2.x>`_ if you use them.


================================================================================
 v2.3.1_ :date:`(8 Feb 2015)`
================================================================================

- Fix issue that prevented non-deferred service providers referencing Breadcrumbs (`#39`_) by making Breadcrumbs non-deferred also (backported from 3.0.0)

.. _v2.3.1: https://github.com/davejamesmiller/laravel-breadcrumbs/tree/3.0.0


================================================================================
 v2.3.0_ :date:`(26 Oct 2014)`
================================================================================

- Add ``$data`` parameter to ``$breadcrumb->push()`` to allow for arbitrary data (`#34`_, `#35`_, `#55`_, `#56`_)

.. _v2.3.0: https://github.com/davejamesmiller/laravel-breadcrumbs/tree/2.3.0
.. _#34: https://github.com/davejamesmiller/laravel-breadcrumbs/issues/34
.. _#35: https://github.com/davejamesmiller/laravel-breadcrumbs/issues/35
.. _#55: https://github.com/davejamesmiller/laravel-breadcrumbs/pull/55
.. _#56: https://github.com/davejamesmiller/laravel-breadcrumbs/pull/56
.. _3a0afc2: https://github.com/laravel/framework/commit/3a0afc20f25ad3bed640ff1a14957f972d123cf7


================================================================================
 v2.2.3_ :date:`(10 Sep 2014)`
================================================================================

- Fix ``Breadcrumbs::generate()`` with no parameters so it uses the current route, like ``Breadcrumbs::render()`` does (`#46`_)

.. _v2.2.3: https://github.com/davejamesmiller/laravel-breadcrumbs/tree/2.2.3
.. _#46: https://github.com/davejamesmiller/laravel-breadcrumbs/issues/46


================================================================================
 v2.2.2_ :date:`(3 Aug 2014)`
================================================================================

- Support for Laravel's ``App::missing()`` method when using automatic route detection (`#40`_, `#41`_)

.. _v2.2.2: https://github.com/davejamesmiller/laravel-breadcrumbs/tree/2.2.2
.. _#40: https://github.com/davejamesmiller/laravel-breadcrumbs/issues/40
.. _#41: https://github.com/davejamesmiller/laravel-breadcrumbs/pull/41


================================================================================
 v2.2.1_ :date:`(19 May 2014)`

src/ServiceProvider.php
<?php namespace Laracraft\Breadcrumbs;

use Illuminate\Container\Container;
use Illuminate\Support\ServiceProvider as BaseServiceProvider;

class ServiceProvider extends BaseServiceProvider {

	/**
	 * Indicates if loading of the provider is deferred.
	 *
	 * @var bool
	 */
	// Can't enable this because there appears to be a bug in Laravel where a
	// non-deferred service provider can't use a deferred one because the boot
	// method is not called - see DependantServiceProviderTest.
	// protected $defer = true;

	/**
	 * Get the services provided by the provider.
	 *
	 * @return array
	 */
	public function provides()
	{
		return ['breadcrumbs'];
	}

	/**
	 * Register the service provider.
	 *
	 * @return void
	 */
	public function register()
	{
	    $this->app->bind('Laracraft\Breadcrumbs\Generator', function ($app) {
	        return new Generator(new Container(), $app['config']['breadcrumbs.namespace']);
        });

		$this->app['breadcrumbs'] = $this->app->share(function($app)
		{
			$breadcrumbs = $this->app->make('Laracraft\Breadcrumbs\Manager');

			$viewPath = __DIR__ . '/../views/';

			$this->loadViewsFrom($viewPath, 'breadcrumbs');
			$breadcrumbs->setView($app['config']['breadcrumbs.view']);

			return $breadcrumbs;
		});
	}

	/**
	 * Bootstrap the application events.
	 *
	 * @return void
	 */
	public function boot()
	{
		$configFile = __DIR__ . '/../config/breadcrumbs.php';

		$this->mergeConfigFrom($configFile, 'breadcrumbs');

		$this->publishes([
			$configFile => config_path('breadcrumbs.php')
		]);

		$this->registerBreadcrumbs();
		$this->registerRoutesBreadcrumbs();
	}

	// This method can be overridden in a child class
	public function registerBreadcrumbs()
	{
		// Load the app breadcrumbs if they're in routes/breadcrumbs.php (Laravel 5.3)
		if (file_exists($file = $this->app['path.base'].'/routes/breadcrumbs.php'))
		{
			require $file;
		}

		// Load the app breadcrumbs if they're in app/Http/breadcrumbs.php (Laravel 5.0-5.2)
		elseif (file_exists($file = $this->app['path'].'/Http/breadcrumbs.php'))
		{
			require $file;
		}
	}

    private function registerRoutesBreadcrumbs()
    {
        $routes = new RouteCollection($this->app['router']->getRoutes());

        $this->app['breadcrumbs']->registerFromRoutes($routes);
    }

}
